cpp/src/arrow/python/numpy_convert.cc

```diff
@@ -274,5 +275,348 @@ Status TensorToNdarray(const std::shared_ptr<Tensor>& tensor, PyObject* base,
+  void* mutable_indptr = const_cast<void*>(immutable_indptr);
+  void* mutable_indices = const_cast<void*>(immutable_indices);
+
+  int array_flags = 0;
+  if (sparse_tensor->is_mutable()) {
+    array_flags |= NPY_ARRAY_WRITEABLE;
+  }
+
+  int type_num_indptr;
+  RETURN_NOT_OK(GetNumPyType(*sparse_index_indptr->type(), &type_num_indptr));
+  PyArray_Descr* dtype_indptr = PyArray_DescrNewFromType(type_num_indptr);
+
+  int type_num_indices;
+  RETURN_NOT_OK(GetNumPyType(*sparse_index_indptr->type(), &type_num_indices));
+  PyArray_Descr* dtype_indices = PyArray_DescrNewFromType(type_num_indices);
+
+  PyObject* result_data =
+      PyArray_NewFromDescr(&PyArray_Type, dtype_data, 1, npy_shape_data.data(), nullptr,
+                           mutable_data, array_flags, nullptr);
+  PyObject* result_indptr =
+      PyArray_NewFromDescr(&PyArray_Type, dtype_indptr, 1, npy_shape_indptr.data(),
+                           nullptr, mutable_indptr, array_flags, nullptr);
+  PyObject* result_indices =
+      PyArray_NewFromDescr(&PyArray_Type, dtype_indices, 1, npy_shape_indices.data(),
+                           nullptr, mutable_indices, array_flags, nullptr);
+  RETURN_IF_PYERROR()
+
+  if (base == Py_None || base == nullptr) {
+    base = py::wrap_sparse_tensor_csr(sparse_tensor);
+  } else {
+    Py_XINCREF(base);
+  }
+  PyArray_SetBaseObject(reinterpret_cast<PyArrayObject*>(result_data), base);
+  PyArray_SetBaseObject(reinterpret_cast<PyArrayObject*>(result_indptr), base);
+  PyArray_SetBaseObject(reinterpret_cast<PyArrayObject*>(result_indices), base);
+
+  *out_data = result_data;
+  *out_indptr = result_indptr;
+  *out_indices = result_indices;
+  return Status::OK();
+}
+
+Status NdarrayToSparseTensorCOO(MemoryPool* pool, PyObject* ao,
+                                std::shared_ptr<SparseTensorCOO>* out) {
+  if (!PyArray_Check(ao)) {
+    return Status::TypeError("Did not pass ndarray object");
+  }
+
+  PyArrayObject* ndarray = reinterpret_cast<PyArrayObject*>(ao);
+
+  // TODO(wesm): What do we want to do with non-contiguous memory and negative strides?
+
+  int ndim = PyArray_NDIM(ndarray);
+
+  // This is also holding the GIL, so don't already draw it.
+  std::shared_ptr<Buffer> data = std::make_shared<NumPyBuffer>(ao);
+  std::vector<int64_t> shape(ndim);
+  std::vector<int64_t> strides(ndim);
+
+  {
+    PyAcquireGIL lock;
+    npy_intp* array_strides = PyArray_STRIDES(ndarray);
+    npy_intp* array_shape = PyArray_SHAPE(ndarray);
+    for (int i = 0; i < ndim; ++i) {
+      if (array_strides[i] < 0) {
+        return Status::Invalid("Negative ndarray strides not supported");
+      }
+      shape[i] = array_shape[i];
+      strides[i] = array_strides[i];
+    }
+
+    std::shared_ptr<DataType> type;
+    RETURN_NOT_OK(
+        GetTensorType(reinterpret_cast<PyObject*>(PyArray_DESCR(ndarray)), &type));
+    Tensor tensor(type, data, shape, strides);
+    *out = std::make_shared<SparseTensorCOO>(tensor);
```

Comment:
AFAICT, this is the place where you want to release the GIL, right?

Reply:
Got it. Is that done by closing scope or should I add a specific command?